### PR TITLE
Screen-space shadows

### PR DESCRIPTION
**Pull request by arose**
Simple screen-space shadows. Changes default light to be in top-left. Shadows are generally off by default but on for cellpack models. More useful for large scale shadows but with some scene-dependent tweaking OK for small-scale stuff too.

### CODE REVIEW

**Comment by arose** [comment]
image

**Comment by arose** [comment]
image

**Comment by giagitom** [comment]
This is awesome!

File: src/mol-gl/shader/postprocessing.frag.ts [hunk 1]
```diff
@@ -120,7 +121,20 @@ void main(void) {
...
+        if (!isBackground(opaqueDepth)) {
+            viewDist = abs(getViewZ(opaqueDepth));
+            fogFactor = smoothstep(uFogNear, uFogFar, viewDist);
+            vec4 shadow = texture2D(tShadows, coords);
+            if (!uTransparentBackground) {
+                color.rgb = mix(mix(vec3(0), uFogColor, fogFactor), color.rgb, shadow.a);
+            } else {
+                color.rgb = mix(vec3(0) * (1.0 - fogFactor), color.rgb, shadow.a);
+            }
+        }
```
Comment (dsehnal) [inline src/mol-gl/shader/postprocessing.frag.ts:134]: Do you think we could get significant extra performance by optimizing the shaders? E.g. reducing the number of operations that happen across the board.

File: src/mol-gl/shader/shadows.frag.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+uniform mat4 uProjection;
+uniform mat4 uInvProjection;
+
+uniform float uMaxDistance;
+uniform float uTolerance;
+uniform float uBias;
+
+bool isBackground(const in float depth) {
+    return depth == 1.0;
```
Comment (dsehnal) [inline src/mol-gl/shader/shadows.frag.ts:35]: Won't this  break with different precisions/drivers/etc?

**Review by dsehnal — APPROVED** [review]
Very cool!

File: src/mol-gl/shader/shadows.frag.ts (hunk 2 above)
Comment (arose) [reply src/mol-gl/shader/shadows.frag.ts:35]: We use it in a bunch of places and haven't seen issues. The 1.0 comes from clearing the texture to 1.0.

File: src/mol-gl/shader/postprocessing.frag.ts (hunk 1 above)
Comment (arose) [reply src/mol-gl/shader/postprocessing.frag.ts:134]: The whole shadow pp takes <0.2 ms, compared to 2ms for occlusion, which is much slower due to more and incoherent texture fetches. I don't think we are at the point to do sub-millisecond optimizations :)

**Pull request merged**